File: app/utils/dawproject_python/warps.py
from lxml import etree as ET
from classes.timeline import Timeline
from classes.warp import Warp
from classes.timeUnit import TimeUnit


class Warps(Timeline):
    def __init__(self, events=None, content=None, content_time_unit=None, **kwargs):
        super().__init__(**kwargs)  # Inherit from Timeline
        self.events = events if events is not None else []
        self.content = content  # Should be an instance of Timeline or its subclass
        self.content_time_unit = content_time_unit  # Instance of TimeUnit

    def to_xml(self):
        # Create the root element for Warps
        warps_elem = ET.Element("Warps")

        # Append the nested content (e.g., another Timeline)
        if self.content:
            warps_elem.append(self.content.to_xml())

        # Recursively add nested Warp elements
        for warp in self.events:
            warps_elem.append(warp.to_xml())

        # Set contentTimeUnit as an attribute
        if self.content_time_unit:
            warps_elem.set("contentTimeUnit", self.content_time_unit.value)

        return warps_elem

    @classmethod
    def from_xml(cls, element):
        # Parse content (which should be a Timeline or subclass)
        content_elem = element.find("Content")
        content = Timeline.from_xml(content_elem) if content_elem is not None else None

        # Recursively parse nested Warp elements
        events = []
        for warp_elem in element.findall("Warp"):
            events.append(Warp.from_xml(warp_elem))

        # Parse the contentTimeUnit attribute
        content_time_unit = element.get("contentTimeUnit")
        content_time_unit = TimeUnit(content_time_unit) if content_time_unit else None

        return cls(events=events, content=content, content_time_unit=content_time_unit)

File: app/utils/dawproject/timeline/warps.ts
import {
  registerTimeline,
  TimelineRegistry,
} from "../registry/timelineRegistry";
import type { ITrack, IWarps } from "../types";
import { Utility } from "../utility";
import { Timeline } from "./timeline";
import { TimeUnit } from "./timeUnit";
import { Warp } from "./warp";

const warpsFactory = (xmlObject: any): Warps => {
  const instance = new Warps();
  instance.fromXmlObject(xmlObject);
  return instance;
};

@registerTimeline("Warps", warpsFactory)
export class Warps extends Timeline implements IWarps {
  contentTimeUnit: TimeUnit;
  points: Warp[];
  content?: Timeline;

  constructor(
    // Make required fields optional for deserialization, provide defaults
    points?: Warp[],
    contentTimeUnit?: TimeUnit,
    content?: Timeline,
    track?: ITrack,
    timeUnit?: TimeUnit,
    name?: string,
    color?: string,
    comment?: string
  ) {
    super(track, timeUnit, name, color, comment); // Pass relevant args to Timeline constructor
    // Provide default placeholders for required fields
    this.points = points || [];
    this.content = content;
    this.contentTimeUnit = contentTimeUnit || TimeUnit.BEATS; // Default placeholder
  }

  toXmlObject(): any {
    const obj: any = {
      Warps: {
        ...super.toXmlObject(), // populate inherited attributes from Timeline
      },
    };

    // add required attribute
    Utility.addAttribute(obj.Warps, "contentTimeUnit", this, {
      required: true,
    });

    // append the nested content (e.g., another Timeline)
    if (this.content) {
      const contentObj = this.content.toXmlObject();
      const tagName = Object.keys(contentObj)[0];
      obj.Warps.Content = { [tagName]: contentObj[tagName] }; // Wrap in "Content" tag
    }

    // recursively add nested Warp elements
    if (this.points && this.points.length > 0) {
      obj.Warps.Warp = this.points.map(warp => warp.toXmlObject().Warp);
    }

    return obj;
  }

  fromXmlObject(xmlObject: any): this {
    super.fromXmlObject(xmlObject); // populate inherited attributes from Timeline

    // validate and populate required attribute
    Utility.populateAttribute<TimeUnit>(xmlObject, "contentTimeUnit", this, {
      required: true,
      castTo: TimeUnit,
    });

    // handle content if present
    let content: Timeline | undefined;
    if (xmlObject.Content) {
      const contentObj = xmlObject.Content;
      const tagName = Object.keys(contentObj)[0]; // Get the actual content tag name

      content = TimelineRegistry.createTimelineFromXml(
        tagName,
        contentObj[tagName]
      );
      if (!content) {
        console.warn(
          `Skipping deserialization of unknown nested timeline content in Warps: ${tagName}`
        );
      }
    }
    this.content = content;

    // recursively parse nested Warp elements
    const points: Warp[] = []; // Renamed from events
    if (xmlObject.Warp) {
      const warpArray = Array.isArray(xmlObject.Warp)
        ? xmlObject.Warp
        : [xmlObject.Warp];
      warpArray.forEach((warpObj: any) => {
        points.push(new Warp().fromXmlObject(warpObj));
      });
    }
    this.points = points;

    return this;
  }
}
